File: js/singles/class.js
class Rectangle {
    // class variable
    static numberOfRectangles = 0;

    // Private fields
    #width;
    #height;

    // Constructor
    constructor(width, height) {
        this.#width = width;
        this.#height = height;
        Rectangle.numberOfRectangles++;
    }

    // Private method
    #iAmPrivate() {
        console.log("This is a private method");
    }

    // Public method to calculate area
    area() {
        return this.#width * this.#height;
    }

    // Public method to draw the rectangle
    draw() {
        this.#iAmPrivate();
        for (let i = 0; i < this.#height; i++) {
            console.log('*'.repeat(this.#width));
        }
    }

    // Public method to get string representation of the object
    toString() {
        return `Rectangle(width=${this.#width}, height=${this.#height})`;
    }
}

// Create an instance of the class with width and height
const rect1 = new Rectangle(10, 10);
console.log(Rectangle.numberOfRectangles);
const rect2 = new Rectangle(10, 20);
console.log(Rectangle.numberOfRectangles);

// Call methods
rect1.draw();
console.log(rect1.area());
console.log(rect1.toString());

File: py/singles/class.py
# Python always need self to access 
# instance attributes and methods
class Rectangle:
    # Constructor/initializer
    # __...__ marks special methods
    def __init__(self, width, height):
        # but __ prefix marks private 
        self.__width = width
        self.__height = height

    # __ marks private
    def __i_am_private(self) :
        print("private")

    # Method to calculate area
    def area(self):
        return self.__width * self.__height

    # Method to draw the rectangle
    def draw(self):
        self.__i_am_private()
        for i in range(self.__height):
            print('*' * self.__width)

    # String representation of the object
    def __str__(self):
        return f"R(width={self.__width}, height={self.__height})"

# Create an instance of the class with width and height
rect = Rectangle(10, 10)
# Call methods
rect.draw()
print(rect.area())
print(rect)
#rect.__i_am_private() does not work
